src/wallet/wallet.cpp

```diff
@@ -857,80 +858,128 @@ bool CWallet::LoadUnifiedAddressMetadata(const ZcashdUnifiedAddressMetadata &add
                 addrmeta.GetReceiverTypes());
 }
 
-PaymentAddress CWallet::GetPaymentAddressForRecipient(
+std::pair<PaymentAddress, RecipientType> CWallet::GetPaymentAddressForRecipient(
         const uint256& txid,
         const libzcash::RecipientAddress& recipient) const
 {
+    AssertLockHeld(cs_main);
+    AssertLockHeld(cs_wallet);
+
     auto self = this;
-    auto defaultAddress = [&]() -> PaymentAddress {
-        auto ufvk = self->GetUFVKForReceiver(RecipientAddressToReceiver(recipient));
-        return std::visit(match {
-            [&](const CKeyID& addr) {
-                auto ua = self->FindUnifiedAddressByReceiver(addr);
-                if (ua.has_value()) {
-                    return libzcash::PaymentAddress{ua.value()};
-                } else {
-                    return libzcash::PaymentAddress{addr};
-                }
-            },
-            [&](const CScriptID& addr) {
-                auto ua = self->FindUnifiedAddressByReceiver(addr);
-                if (ua.has_value()) {
-                    return libzcash::PaymentAddress{ua.value()};
-                } else {
-                    return libzcash::PaymentAddress{addr};
-                }
-            },
-            [&](const SaplingPaymentAddress& addr) {
-                auto ua = self->FindUnifiedAddressByReceiver(addr);
-                if (ua.has_value()) {
-                    return libzcash::PaymentAddress{ua.value()};
-                } else if (ufvk.has_value() && ufvk->GetSaplingKey().has_value()) {
-                    auto saplingKey = ufvk->GetSaplingKey().value();
-                    auto j = saplingKey.DecryptDiversifier(addr.d);
-                    // std::get is safe here because we know we have a valid Sapling diversifier index
-                    auto defaultUA = std::get<std::pair<UnifiedAddress, diversifier_index_t>>(
-                            ufvk->Address(j, CWallet::DefaultReceiverTypes()));
-                    return libzcash::PaymentAddress{defaultUA.first};
-                } else {
-                    return libzcash::PaymentAddress{addr};
+
+    int nHeight = chainActive.Height();
+    auto wtxPtr = mapWallet.find(txid);
+    if (wtxPtr != mapWallet.end()) {
+        const CBlockIndex* pTxIndex{nullptr};
+        if (wtxPtr->second.GetDepthInMainChain(pTxIndex) > 0) {
+            nHeight = pTxIndex->nHeight;
+        }
+    }
+
+    auto ufvk = self->GetUFVKForReceiver(RecipientAddressToReceiver(recipient));
+    std::pair<PaymentAddress, RecipientType> defaultAddress = std::visit(match {
+        [&](const CKeyID& addr) {
```

Comment:
This and the `CScriptID` case are dead code, because `CWallet::GetPaymentAddressForRecipient` is only called from `z_viewtransaction` which does not show transparent information. I argue that we can't get the semantics of these cases correct in time for this PR to merge and that we should just delete these cases. @nuttycom says he will fix this during the RC period. Merging of this PR is blocked on there being an issue blocking 4.7.0 to address this.

Reply:
Fixed in https://github.com/zcash/zcash/pull/5715/commits/151074cbb13455afb8257354c0d856a79da3f332